doc/manual/contract.md

```diff
@@ -0,0 +1,535 @@
+
+The invariant is not checked when the call is on `self`. The following example shows the execution of invariant:
+
+~~~
+class A
+	invariant(print_invariant)
+
+	fun print_invariant: Bool
+	do
+		print "In invariant verification"
+		return true
+	end
+
+	fun foo
+	do
+		print "In foo"
+		self.bar # The called property is a member of "self", the invariant will not be checked.
+	end
+
+	fun bar
+	do
+		print "In bar"
+	end
+end
+
+var a = new A
+a.foo # The called property is not a member of "self", the invariant will be checked.
+~~~
+
+This example gives the following output:
+~~~nitish
+In invariant verification
+In foo
+In bar
+In invariant verification
+~~~
+
+## Writing contracts
+
+Methods, as well as attribute setters, can be associated with two types of contracts; precondition (`expect`) that expresses requirements of the routine´s call and postcondition (`ensure`) that expresses what the routine guarantee after her execution. Together, they define the rules to be respected between the callee and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfill the `ensure` condition on exit.
+
+All contracts (`expect`, `ensure`) are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
+
+When defining a contract (`expect`, `ensure`) on an attribute like bellow, only the automatically generated write property (setter) will use the contract.
+
+~~~nitish
+var baz: Int is expect(baz > 0) # In the contract, baz refers to the parameter of the write property (setter).
+~~~
+
+### Define preconditions with `expect`
+
+We define a precondition using the annotation `expect`. In the `Stack` class, the method `pop` defines a precondition, that a call to the method should be made on a non-empty stack.
+
+~~~nitish
+fun pop is expect(not_empty)
+~~~
+
+### Define postconditions with `ensure`
+
+We define a postcondition using the annotation `ensure`. In the `Stack` class, the method `push` defines postconditions. The first guarantee defines that, the method must add the element received as argument (`object`) at the top of the stack.
+
+~~~nitish
+fun push(object: T) is ensure(self[size - 1] == object)
+~~~
+
+Within an `ensure` contract we can refer to the return value of the method with the keyword `result`. In the `Stack` class, the `[]` method guarantees that, the return value is the same as the element stored in the stack at the given index.
+
+~~~nitish
+fun [](index: Int): T is ensure(result == data[index])
+~~~
+
+Note that the keyword `result` can only be used when the method has a return parameter.
+
+
+In `ensure` we can refer to any arguments and instance variables before the method call with the using of `old` keyword. Again, with the `pop` method, we can ensure that the `size` of the stack after the method call will be the same as before minus one `old(size) - 1 == size`.
```

Comment:
Phrase bizarre vers la fin: before the method call with the using of old
->
In an ensure expression, we can refer to any arguments and instance variables before the method call, using the `old` keyword.